polynote-server/src/main/scala/polynote/server/SharedNotebook.scala

```diff
@@ -214,11 +216,12 @@ class IOSharedNotebook(
       }
     }(_ => kernelLock.release)
 
+    // we want to queue all the cells, but evaluate them in order. So the outer IO of the result runs the outer IO of queueCell for all the cells.
     override def runCells(ids: List[CellID]): IO[Stream[IO, CellResult]] =
-      ensureKernel().map {
-        kernel => Stream.emits(ids).evalMap {
-          id => runCell(id).map(results => results.map(result => CellResult(ShortString(path), id, result)))
-        }.flatten
+      ids.map {
+        id => queueCell(id).map(_.map(_.map(result => CellResult(ShortString(path), id, result))))
+      }.sequence.map {
```

Comment:
Not sure whether this is useful at all, but as I'm looking at the docs right now I'll point out that they say `In general t.map(f).sequence can be replaced with t.traverse(f).`

Reply:
tbh I think `t.map(f).sequence` is clearer

Reply:
Yeah, I think map and sequence is clearer in its intent... `traverse`'s type signature takes some squinting before it makes sense